compiler/optimizer/VPHandlers.cpp

```diff
@@ -5400,10 +5400,75 @@ static void devirtualizeCall(OMR::ValuePropagation *vp, TR::Node *node)
    vp->invalidateValueNumberInfo();
    }
 
+static bool canFoldNonOverriddenGuard(OMR::ValuePropagation *vp, TR::Node *callNode, TR::Node *guardNode)
+   {
+   TR::SymbolReference *symRef        = callNode->getSymbolReference();
+   TR::MethodSymbol    *methodSymbol  = symRef->getSymbol()->castToMethodSymbol();
+
+   int32_t firstArgIndex = callNode->getFirstArgumentIndex();
+   bool isGlobal;
+   TR::VPConstraint *constraint = vp->getConstraint(callNode->getChild(firstArgIndex), isGlobal);
+   TR_OpaqueClassBlock *thisType;
+   dumpOptDetails(vp->comp(), "Guard %p Call %p constraint %p\n", guardNode, callNode, constraint);
+   if (constraint &&
+       constraint->isFixedClass() &&
+       (thisType = constraint->getClass()) &&
+       methodSymbol->isVirtual())
+      {
+      TR::ResolvedMethodSymbol * resolvedMethodSymbol = methodSymbol->getResolvedMethodSymbol();
+      if (resolvedMethodSymbol)
+         {
+         TR_ResolvedMethod *originalResolvedMethod = resolvedMethodSymbol->getResolvedMethod();
+         TR_OpaqueClassBlock *originalMethodClass = originalResolvedMethod->classOfMethod();
+
+         if ((vp->fe()->isInstanceOf(thisType, originalMethodClass, true, true) == TR_yes) &&
+             !originalResolvedMethod->virtualMethodIsOverridden())
+            {
+            TR_VirtualGuard *virtualGuard = vp->comp()->findVirtualGuardInfo(guardNode);
+            if (virtualGuard && virtualGuard->canBeRemoved())
+               return true;
+            }
+         }
+      }
+      return false;      
+   }
+
+
 TR::Node *constrainCall(OMR::ValuePropagation *vp, TR::Node *node)
    {
    constrainChildren(vp, node);
 
+   if (vp->lastTimeThrough() &&
+       vp->_isGlobalPropagation &&
+       (*(vp->_callNodeToGuardNodes))[node])
```

Comment:
Looking up `node` using `operator[]` will insert a null for each absent node queried here. Thankfully it's only the call nodes, but if you want to avoid the insertion, you can use `find()` and compare the result to `end()`. If you keep the result of `find`, that can also eliminate the need for the redundant lookup just below

Reply:
Done